common/lib/core-interfaces/test-d/index.test-d.ts

```diff
@@ -0,0 +1,63 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+/* eslint-disable unicorn/filename-case */
+// eslint-disable-next-line import/no-extraneous-dependencies
+import {expectError} from "tsd";
+import { IFluidLoadable, IProvideFluidLoadable, Provider, ProviderKeys } from "../dist";
+
+
+declare function getUnknown(): Provider;
+
+declare function useUnknown(params: Provider | undefined): void;
+
+declare function useProvider<T extends Provider>(params: Provider<T> | undefined): void;
+
+declare function useProviderKey<T,TKey extends ProviderKeys<T> = ProviderKeys<T>>(key: TKey): void;
+
+declare function useLoadable(params: Provider<IFluidLoadable> | undefined): void;
+
+// test with provider
+{
+    const provider: Provider<IProvideFluidLoadable> = getUnknown();
+    useUnknown(provider);
+    useUnknown(provider.IFluidLoadable);
+    useProvider(provider);
+    useProvider(provider.IFluidLoadable);
+    useLoadable(provider);
+    useLoadable(provider.IFluidLoadable);
+    expectError(provider.handle);
```

Comment:
I don not like this part (.handle & IFluidLoadable loading being on same object).
To me - provider and implementer are two different concepts that should not be mixed (on consuming side) into single object / interface (i.e. IFluidLoadable should not extend IProvideFluidLoadable, and any mechanism we implement should not expose IProvideFluidLoadable & IFluidLoadable as if it were one and the same). Whoever implements IProvideFluidLoadable is free to implement both interfaces, but consumer should not assume that, allowing provider to delegate capabilities of other objects.

So when I see reference to IProvideFoo.IFoo.IFoo.IFoo above, I cringe

Here, I like IUnknonw.QueryInterface and aggregation rules - it is free to return you different objects, but there are some rules RE asking for same interface (you should get same object no matter what object from aggregations was asked for it)

Reply:
I don't disagree with this entirely, but i don't have a solution, and think it is out of scope for the work i'm doing here. I'm really just trying to decouple the typing rats nest we get into via module augmentation and IFluidObject. In general, i'd also like to reduce our reliance on the pattern, as it should really only be used for feature discover on anonymous types. Ideally, we can get the runtime to a state where this type of pattern is not central, and we can let consumers bring their own pattern here if they need one.

specifically, I'm solving:
#2566
 #3264